lib/concurrent/atomic/thread_local_var.rb

```diff
@@ -115,6 +115,104 @@ def set(value)
     end
   end
 
+  require 'monitor'
+  class AlexThreadLocalVar < AbstractThreadLocalVar
+
+    # Each thread has a (lazily initialized) array of thread-local variable values
+    # Each time a new thread-local var is created, we allocate an "index" for it
+    # For example, if the allocated index is 1, that means slot #1 in EVERY
+    #   thread's thread-local array will be used for the value of that TLV
+    #
+    # The good thing about using a per-THREAD structure to hold values, rather
+    #   than a per-TLV structure, is that no synchronization is needed when
+    #    reading and writing those values (since the structure is only ever
+    #    accessed by a single thread)
+    #
+    # Of course, when a TLV is GC'd, 1) we need to recover its index for use
+    #   by other new TLVs (otherwise the thread-local arrays could get bigger
+    #   and bigger with time), and 2) we need to null out all the references
+    #   held in the now-unused slots (both to avoid blocking GC of those objects,
+    #   and also to prevent "stale" values from being passed on to a new TLV
+    #   when the index is reused)
+    # Because we need to null out freed slots, we need to keep references to
+    #   ALL the thread-local arrays -- ARRAYS is for that
+    # But when a Thread is GC'd, we need to drop the reference to its thread-local
+    #   array, so we don't leak memory
+
+    FREE = []
+    LOCK = Monitor.new
+    ARRAYS = {} # used as a hash set
+    @@next = 0
+
+    protected
+
+    def self.threadlocal_finalizer(index)
+      lambda do
+        LOCK.synchronize do
+          FREE.push(index)
+          # The cost of GC'ing a TLV is linear in the number of threads using TLVs
+          # But that is natural! More threads means more storage is used per TLV
+          # So naturally more CPU time is required to free more storage
+          ARRAYS.each_key do |array|
+            array[index] = nil
```

Comment:
Aha, nice catch. So I really want something like ARRAYS[array.object_id] --
in other words, this was intended to be an identity set.
On Jul 22, 2015 5:00 AM, "thedarkone" <notifications@github.com> wrote:

> In lib/concurrent/atomic/thread_local_var.rb
> <https://github.com/ruby-concurrency/concurrent-ruby/pull/376#discussion_r35177594>
> :
>
> > +    FREE = []
> > +    LOCK = Monitor.new
> > +    ARRAYS = {} # used as a hash set
> > +    @@next = 0
> > +
> > +    protected
> > +
> > +    def self.threadlocal_finalizer(index)
> > +      lambda do
> > +        LOCK.synchronize do
> > +          FREE.push(index)
> > +          # The cost of GC'ing a TLV is linear in the number of threads using TLVs
> > +          # But that is natural! More threads means more storage is used per TLV
> > +          # So naturally more CPU time is required to free more storage
> > +          ARRAYS.each_key do |array|
> > +            array[index] = nil
>
> You're changing a Hash key (array) here, its hash value might change, so
> after the each_key loop you need to "rehash the hash": ARRAY.rehash.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ruby-concurrency/concurrent-ruby/pull/376/files#r35177594>
> .
>